File: app/src/main/java/srjhlab/com/myownbarcode/CommonUi/AddBarcodeInfoDialog.java
package srjhlab.com.myownbarcode.CommonUi;

import android.app.DialogFragment;
import android.databinding.DataBindingUtil;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;

import org.apache.commons.lang.StringUtils;
import org.greenrobot.eventbus.EventBus;

import srjhlab.com.myownbarcode.Item.BarcodeItem;
import srjhlab.com.myownbarcode.R;
import srjhlab.com.myownbarcode.Utils.CommonEventbusObejct;
import srjhlab.com.myownbarcode.Utils.CommonUtils;
import srjhlab.com.myownbarcode.Utils.ConstVariables;
import srjhlab.com.myownbarcode.Utils.MakeBarcode;
import srjhlab.com.myownbarcode.Utils.ToastUtil;
import srjhlab.com.myownbarcode.databinding.FragmentAddinfoBinding;

public class AddBarcodeInfoDialog extends DialogFragment implements View.OnClickListener {
    private final static String TAG = AddBarcodeInfoDialog.class.getSimpleName();
    public final static int MODE_ADD_BARCODE = 0;
    public final static int MODE_EDIT_BARCODE = 1;

    private FragmentAddinfoBinding mBinding;
    private BarcodeItem mItem = null;

    private int mPicColor = -1;
    private Drawable mDrawsable = null;
    private Bitmap mBitmap = null;
    private int mCommandType;

    public static AddBarcodeInfoDialog newInstance() {
        return new AddBarcodeInfoDialog();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        Log.d(TAG, "##### onCreateVIew #####");
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_addinfo, container, false);

        getDialog().getWindow().getAttributes().windowAnimations = R.style.SelectDialogAnimation;
        getDialog().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        getDialog().setCanceledOnTouchOutside(true);

        initializeUI();
        return mBinding.getRoot();
    }

    @Override
    public void dismiss() {
        Log.d(TAG, "##### dismiss #####");
        super.dismiss();
    }

    private void initializeUI() {
        Log.d(TAG, "##### initializeUI #####");

        if (mItem != null) {
            mBinding.textviewTypeDialogAddBarcode.setText(CommonUtils.convertBarcodeType(getActivity(), mItem.getBarcodeType()));
            String strArr[] = CommonUtils.splitStringEvery(mItem.getBarcodeValue(), 4);
            String str = StringUtils.join(strArr, " ");
            mBinding.textviewValueDialogAddBarcode.setText(str);
        }

        mBinding.imageviewOkDialogAddBarcode.setOnClickListener(this);
        mBinding.imageviewCancelDialogAddBarcode.setOnClickListener(this);
        mBinding.colorPic1.setOnClickListener(this);
        mBinding.colorPic2.setOnClickListener(this);
        mBinding.colorPic3.setOnClickListener(this);
        mBinding.colorPic4.setOnClickListener(this);
        mBinding.colorPic5.setOnClickListener(this);
        mBinding.colorPic6.setOnClickListener(this);
        mBinding.colorPic7.setOnClickListener(this);
        mBinding.colorPic8.setOnClickListener(this);
        mBinding.colorPic9.setOnClickListener(this);
        mBinding.colorPic10.setOnClickListener(this);
        mBinding.colorPic11.setOnClickListener(this);
        mBinding.colorPic12.setOnClickListener(this);

        if (mItem.getBarcodeValue() != null) {
            setOverviewBarcode(mItem.getBarcodeType(), mItem.getBarcodeValue());
        }

        if (mCommandType == MODE_EDIT_BARCODE) {
            if (mItem != null) {
                mBinding.edittextDialogAddBarcode.setText(mItem.getBarcodeName());
                setPreSelecColor(mItem.getBarcodeCardColor());
            }
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.imageview_ok_dialog_add_barcode:
                Log.d(TAG, "##### onCLick ##### imageview_ok_dialog_add_barcode");
                saveBarcode();
                break;
            case R.id.imageview_cancel_dialog_add_barcode:
                Log.d(TAG, "##### onCLick ##### imageview_cancel_dialog_add_barcode");
                dismiss();
                break;
            case R.id.color_pic_1:
                Log.d(TAG, "##### onCLick ##### color_pic_1");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_1);
                mBinding.colorPic1.setSelected(true);
                break;
            case R.id.color_pic_2:
                Log.d(TAG, "##### onCLick ##### color_pic_2");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_2);
                mBinding.colorPic2.setSelected(true);
                break;
            case R.id.color_pic_3:
                Log.d(TAG, "##### onCLick ##### color_pic_3");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_3);
                mBinding.colorPic3.setSelected(true);
                break;
            case R.id.color_pic_4:
                Log.d(TAG, "##### onCLick ##### color_pic_4");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_4);
                mBinding.colorPic4.setSelected(true);
                break;
            case R.id.color_pic_5:
                Log.d(TAG, "##### onCLick ##### color_pic_5");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_5);
                mBinding.colorPic5.setSelected(true);
                break;
            case R.id.color_pic_6:
                Log.d(TAG, "##### onCLick ##### color_pic_6");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_6);
                mBinding.colorPic6.setSelected(true);
                break;
            case R.id.color_pic_7:
                Log.d(TAG, "##### onCLick ##### color_pic_7");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_7);
                mBinding.colorPic7.setSelected(true);
                break;
            case R.id.color_pic_8:
                Log.d(TAG, "##### onCLick ##### color_pic_8");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_8);
                mBinding.colorPic8.setSelected(true);
                break;
            case R.id.color_pic_9:
                Log.d(TAG, "##### onCLick ##### color_pic_9");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_9);
                mBinding.colorPic9.setSelected(true);
                break;
            case R.id.color_pic_10:
                Log.d(TAG, "##### onCLick ##### color_pic_10");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_10);
                mBinding.colorPic10.setSelected(true);
                break;
            case R.id.color_pic_11:
                Log.d(TAG, "##### onCLick ##### color_pic_11");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_11);
                mBinding.colorPic11.setSelected(true);
                break;
            case R.id.color_pic_12:
                Log.d(TAG, "##### onCLick ##### color_pic_12");
                clearSelectedColor();
                mPicColor = getResources().getColor(R.color.color_pic_12);
                mBinding.colorPic12.setSelected(true);
                break;
        }
    }

    private void saveBarcode() {
        Log.d(TAG, "##### saveBarcode #####");

        if (mPicColor == -1) {
            ToastUtil.getInstance(getActivity()).showToast(getResources().getString(R.string.string_request_pic_color));
            return;
        }

        if (mBinding.edittextDialogAddBarcode.getText().toString() == null
                && mBinding.edittextDialogAddBarcode.getText().toString() == "") {
            ToastUtil.getInstance(getActivity()).showToast(getResources().getString(R.string.string_request_write_name));
            return;
        }

        //    public BarcodeItem(int type, int barcodeType, int barcodeId, String barcodeName, int barcodeColor, String barcodeValue, Bitmap barcodeBitmap) {
        if (mCommandType == MODE_ADD_BARCODE) {
            //mDbHelper.insert(mBinding.edittextDialogAddBarcode.getText().toString(), mPicColor, mItem.getBarcodeValue(), mDrawsable);
            EventBus.getDefault().post(new CommonEventbusObejct(ConstVariables.EVENTBUS_ADD_NEW_BARCODE, new BarcodeItem(ConstVariables.ITEM_TYPE_BARCODE, mItem.getBarcodeType(), mItem.getBarcodeId(), mBinding.edittextDialogAddBarcode.getText().toString(), mPicColor, mItem.getBarcodeValue(), mBitmap)));
        } else if (mCommandType == MODE_EDIT_BARCODE) {
            //mDbHelper.update(mItem.getBarcodeId(), mBinding.edittextDialogAddBarcode.getText().toString(), mPicColor);
            EventBus.getDefault().post(new CommonEventbusObejct(ConstVariables.EVENTBUS_MODIFY_BARCODE, new BarcodeItem(ConstVariables.ITEM_TYPE_BARCODE, mItem.getBarcodeType(), mItem.getBarcodeId(), mBinding.edittextDialogAddBarcode.getText().toString(), mPicColor, mItem.getBarcodeValue(), mItem.getBarcodeBitmap())));
        }
        dismiss();
    }

    private void setPreSelecColor(int color) {
        Log.d(TAG, "##### setPreSelectColor ##### color : " + color);
        if(getActivity() == null){
            return;
        }
        if(color == getActivity().getColor(R.color.color_pic_1)){
            mBinding.colorPic1.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_2)){
            mBinding.colorPic2.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_3)){
            mBinding.colorPic3.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_4)){
            mBinding.colorPic4.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_5)){
            mBinding.colorPic5.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_6)){
            mBinding.colorPic6.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_7)){
            mBinding.colorPic7.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_8)){
            mBinding.colorPic8.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_9)){
            mBinding.colorPic9.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_10)){
            mBinding.colorPic10.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_11)){
            mBinding.colorPic11.setSelected(true);
        }else if(color == getActivity().getColor(R.color.color_pic_12)){
            mBinding.colorPic12.setSelected(true);
        }
    }

    private void clearSelectedColor() {
        Log.d(TAG, "##### clearSelectColor #####");

        mBinding.colorPic1.setSelected(false);
        mBinding.colorPic2.setSelected(false);
        mBinding.colorPic3.setSelected(false);
        mBinding.colorPic4.setSelected(false);
        mBinding.colorPic5.setSelected(false);
        mBinding.colorPic6.setSelected(false);
        mBinding.colorPic7.setSelected(false);
        mBinding.colorPic8.setSelected(false);
        mBinding.colorPic9.setSelected(false);
        mBinding.colorPic10.setSelected(false);
        mBinding.colorPic11.setSelected(false);
        mBinding.colorPic12.setSelected(false);
    }

    public AddBarcodeInfoDialog setCommandType(int type) {
        Log.d(TAG, "##### setCommandType #####");
        this.mCommandType = type;
        return this;
    }

    public AddBarcodeInfoDialog setBarcodeItem(BarcodeItem item) {
        Log.d(TAG, "##### setBarcodeItem ##### type : " + item.getBarcodeType() + " value : " + item.getBarcodeValue());
        this.mItem = item;
        return this;
    }

    private void setOverviewBarcode(int type, String value) {
        Log.d(TAG, "##### setOverviewBarcode #####");
        Bitmap imageBitmap = MakeBarcode.getInstance().makeBarcode(type, value);
        mBitmap = imageBitmap;
        mBinding.imageviewBarcodeDialogAddBarcode.setImageBitmap(imageBitmap);
        mDrawsable = new BitmapDrawable(imageBitmap);
    }
}

File: app/src/main/kotlin/Dialog/AddBarcodeInfoDialog.kt
package srjhlab.com.myownbarcode.Dialog

import android.app.DialogFragment
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.ImageView
import kotlinx.android.synthetic.main.layout_dialog_addinfo.*
import org.apache.commons.lang.StringUtils
import org.greenrobot.eventbus.EventBus
import org.jetbrains.anko.toast
import srjhlab.com.myownbarcode.Item.BarcodeItem
import srjhlab.com.myownbarcode.R
import srjhlab.com.myownbarcode.Utils.CommonEventbusObejct
import srjhlab.com.myownbarcode.Utils.CommonUtils
import srjhlab.com.myownbarcode.Utils.ConstVariables
import srjhlab.com.myownbarcode.Utils.MakeBarcode

@Suppress("DEPRECATION")

class AddBarcodeInfoDialog : DialogFragment(), View.OnClickListener {
    private val TAG = this.javaClass.simpleName

    companion object {
        const val MODE_ADD_BARCODE = 0
        const val MODE_EDIT_BARCODE = 1
    }

    private lateinit var mItem: BarcodeItem
    private var mCommandType: Int = -1
    private lateinit var mBitmap: Bitmap
    private lateinit var mDrawable: Drawable
    private lateinit var mColorArr: Array<Int>
    private lateinit var mColorPickerArr: Array<ImageView>
    private var mPicColor: Int = -1

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        Log.d(TAG, "##### onCreateView #####")
        dialog.window.attributes.windowAnimations = R.style.SelectDialogAnimation
        dialog.window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN)
        dialog.window.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog.setCanceledOnTouchOutside(true)

        mColorArr = arrayOf(
                activity.getColor(R.color.color_pic_1)
                , activity.getColor(R.color.color_pic_2)
                , activity.getColor(R.color.color_pic_3)
                , activity.getColor(R.color.color_pic_4)
                , activity.getColor(R.color.color_pic_5)
                , activity.getColor(R.color.color_pic_6)
                , activity.getColor(R.color.color_pic_7)
                , activity.getColor(R.color.color_pic_8)
                , activity.getColor(R.color.color_pic_9)
                , activity.getColor(R.color.color_pic_10)
                , activity.getColor(R.color.color_pic_11)
                , activity.getColor(R.color.color_pic_12)
        )
        return inflater.inflate(R.layout.layout_dialog_addinfo, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        Log.d(TAG, "##### onACtivityCReated #####")
        initializeUI()
    }

    override fun onClick(v: View) {
        Log.d(TAG, "##### onClick ##### v.id : ${v.id}")
        when (v.id) {
            textview_dialog_ok.id -> {
                saveBarcode()
                dismiss()
                return
            }
            textview_dialog_cancel.id -> {
                dismiss()
                return
            }
        }
        for ((index, i) in mColorArr.withIndex()) {
            if (mColorPickerArr[index].id.equals(v.id)) {
                clearSelectedColor()
                this.mPicColor = mColorArr[index]
                mColorPickerArr[index].isSelected = true
                return
            }
        }
    }

    fun initializeUI() {
        Log.d(TAG, "##### initializeUI #####")
        textview_dialog_ok.setOnClickListener(this)
        textview_dialog_cancel.setOnClickListener(this)

        mColorPickerArr = arrayOf(
                color_pic_1
                , color_pic_2
                , color_pic_3
                , color_pic_4
                , color_pic_5
                , color_pic_6
                , color_pic_7
                , color_pic_8
                , color_pic_9
                , color_pic_10
                , color_pic_11
                , color_pic_12)

        for ((index, i) in mColorPickerArr.withIndex()) {
            mColorPickerArr[index].setOnClickListener(this)
        }

        val strArr: Array<String> = CommonUtils.splitStringEvery(mItem.barcodeValue, 4)
        val str = StringUtils.join(strArr, " ")
        textview_value_dialog_add_barcode.text = str

        if (mItem.barcodeValue != null) {
            setOverviewBarcode(mItem.barcodeType, mItem.barcodeValue)
        }

        if (mItem.barcodeType != null) {
            textview_type_dialog_add_barcode.text = CommonUtils.convertBarcodeType(activity, mItem.barcodeType)
        }

        if (mCommandType == MODE_EDIT_BARCODE) {
            edittext_dialog_add_barcode.setText(mItem.barcodeName)
            setPreSelectedColor(mItem.barcodeCardColor)
            mPicColor = mItem.barcodeCardColor
        }
    }


    fun setCommandType(type: Int): AddBarcodeInfoDialog {
        Log.d(TAG, "##### setCommandType #####")
        this.mCommandType = type
        return this
    }

    fun setBarcodeItem(item: BarcodeItem): AddBarcodeInfoDialog {
        Log.d(TAG, "##### setBarcodeItem #####")
        this.mItem = item
        return this
    }

    private fun saveBarcode() {
        Log.d(TAG, "##### saveBarcode #####")
        if (mPicColor == -1) {
//            ToastUtil.getInstance(activity).showToast(resources.getString(R.string.string_request_pic_color))
            toast(resources.getString(R.string.string_request_pic_color))
            return
        }

        if (edittext_dialog_add_barcode.text.toString() == "") {
//            ToastUtil.getInstance(activity).showToast(resources.getString(R.string.string_request_write_name))
            toast(resources.getString(R.string.string_request_write_name))
            return
        }

        when (mCommandType) {
            MODE_ADD_BARCODE -> EventBus.getDefault().post(CommonEventbusObejct(ConstVariables.EVENTBUS_ADD_NEW_BARCODE, BarcodeItem(ConstVariables.ITEM_TYPE_BARCODE, mItem.barcodeType, mItem.barcodeId, edittext_dialog_add_barcode.text.toString(), mPicColor, mItem.barcodeValue, mBitmap)))
            MODE_EDIT_BARCODE -> EventBus.getDefault().post(CommonEventbusObejct(ConstVariables.EVENTBUS_MODIFY_BARCODE, BarcodeItem(ConstVariables.ITEM_TYPE_BARCODE, mItem.barcodeType, mItem.barcodeId, edittext_dialog_add_barcode.text.toString(), mPicColor, mItem.barcodeValue, mItem.barcodeBitmap)))
        }
        dismiss()

    }

    private fun setPreSelectedColor(color: Int) {
        Log.d(TAG, "##### setPreSelectedColor #####")
        when (color) {
            activity.getColor(R.color.color_pic_1) -> mColorPickerArr.get(0).isSelected = true
            activity.getColor(R.color.color_pic_2) -> mColorPickerArr.get(1).isSelected = true
            activity.getColor(R.color.color_pic_3) -> mColorPickerArr.get(2).isSelected = true
            activity.getColor(R.color.color_pic_4) -> mColorPickerArr.get(3).isSelected = true
            activity.getColor(R.color.color_pic_5) -> mColorPickerArr.get(4).isSelected = true
            activity.getColor(R.color.color_pic_6) -> mColorPickerArr.get(5).isSelected = true
            activity.getColor(R.color.color_pic_7) -> mColorPickerArr.get(6).isSelected = true
            activity.getColor(R.color.color_pic_8) -> mColorPickerArr.get(7).isSelected = true
            activity.getColor(R.color.color_pic_9) -> mColorPickerArr.get(8).isSelected = true
            activity.getColor(R.color.color_pic_10) -> mColorPickerArr.get(9).isSelected = true
            activity.getColor(R.color.color_pic_11) -> mColorPickerArr.get(10).isSelected = true
            activity.getColor(R.color.color_pic_12) -> mColorPickerArr.get(11).isSelected = true
        }
    }

    private fun clearSelectedColor() {
        Log.d(TAG, "#### clearSelectedColor #####")
        for ((index, i) in mColorPickerArr.withIndex()) {
            mColorPickerArr[index].isSelected = false
        }
    }

    private fun setOverviewBarcode(type: Int, value: String) {
        Log.d(TAG, "##### setOverviewBarcode #####")
        var imageBitmap = MakeBarcode.getInstance().makeBarcode(type, value)
        this.mBitmap = imageBitmap
        imageview_barcode_dialog_add_barcode.setImageBitmap(mBitmap)
        this.mDrawable = BitmapDrawable(mBitmap)
    }
}